### ISSUE TITLE
Add authorization error messages to data or extensions instead of as a top level error

### ISSUE THREAD

**Issue by crpahl** [rmosolgo/graphql-ruby#4124]
**Describe the bug**

Sorry, this is definitely more of a question than a bug.

We want to use the default behaviour of having the framework silently replace unauthorized objects with nil, but we would also like to add authorization error messages to the response so the client can have some insight into why a field returned `null`.

As a rough example, something like this works and will simply append the authorization errors to the extensions hash:

```
        def unauthorized_object(error)
          message = "An object of type #{error.type.graphql_name} was hidden due to permissions"
          add_authorization_error(error, message)
        end

        def unauthorized_field(error)
          message = "The field #{error.field.graphql_name} on an object of type #{error.type.graphql_name} was hidden due to permissions"
          add_authorization_error(error, message)
        end

        def add_authorization_error(error, message)
          (error.context.response_extensions[:authorization_errors] ||= []).push(message)
          error.context.response_extensions[:authorization_errors].uniq!
        end
```

Does something like this approach make sense, or is there some other hook, or part of the framework that we should be using to accomplish this?

**Versions**

1.13

**GraphQL schema**

**Comment by rmosolgo** [comment]
Yep, that's the way to do it! I don't know of any better way of handling errors by adding to the response extensions. 

The only thing I'd add is maybe returning `nil` from `def unauthorized_object` and `def unauthorized_field` -- some of those hooks' return values are _used_ by GraphQL-Ruby, others are ignored, and I don't remember which are which 🙈

**Comment by rmosolgo** [comment]
Let me know if you need anything further on this one, but it sounds like that should do the job :+1:
